cumulusci/tasks/bulkdata/delete.py

```diff
@@ -38,77 +41,46 @@ def _init_options(self, kwargs):
                 "Criteria cannot be specified if more than one object is specified."
             )
         self.options["hardDelete"] = process_bool_arg(self.options.get("hardDelete"))
+        self.options["ignore_row_errors"] = process_bool_arg(
+            self.options.get("ignore_row_errors")
+        )
 
     def _run_task(self):
         for obj in self.options["objects"]:
             self.logger.info(f"Deleting {self._object_description(obj)} ")
-            delete_job = self._create_job(obj, self.options["where"])
-            if delete_job is not None:
-                self._wait_for_job(delete_job)
-
-    def _create_job(self, obj, where=None):
-        # Query for rows to delete
-        delete_rows = self._query_salesforce_for_records_to_delete(obj, where)
-        if not delete_rows:
-            self.logger.info(f"  No {obj} objects found, skipping delete")
-            return
 
-        # Upload all the batches
-        operation = "hardDelete" if self.options["hardDelete"] else "delete"
-        delete_job = self.bulk.create_job(obj, operation)
-        self.logger.info(f"  Deleting {len(delete_rows)} {obj} records")
-        batch_num = 1
-        for batch in self._upload_batches(delete_job, delete_rows):
-            self.logger.info(f"    Uploaded batch {batch}")
-            batch_num += 1
-        self.bulk.close_job(delete_job)
-        return delete_job
-
-    def compose_query(self, obj, where):
-        query = f"SELECT Id FROM {obj}"
-        if where:
-            query += f" WHERE {where}"
+            query = f"SELECT Id FROM {obj}"
+            if self.options["where"]:
+                query += f" WHERE {self.options['where']}"
+
+            if not self.options["where"]:
+                # FIXME: Perform a count() query to determine whether we have work to do.
+                pass
+
+            qs = BulkApiQueryStep(obj, {}, self, query)
+            qs.query()
```

Comment:
In my quick testing, qs.status is not set when qs.query runs.

Also, it is unclear to me how the actual query results get fed into the delete step below.

Reply:
I'll check on the status update issue.

The records form query are fed in at line 75. The query's result generator becomes the delete's input generator (through a `map()` to change its structure).

Reply:
Okay I see it now. I was imagining the stream being submitted in the step constructor.

Is it important that it be exposed to the caller as a 4-step process? construct, start, feed, end? 

In the two places its used, the steps seem identical and therefore could perhaps be hidden in a 1-step or 2-step process?

Reply:
I fixed the issue with `status` not being set. It turns out we have all along been silently ignoring failures at the query batch level, which I've fixed.

As to the 4-step process: I'm open to condensing this. Mostly, I think that would look like moving the logic of `start()` and `end()` into `load_records()`. But I liked keeping the option open for `load_records()` to be called multiple times to help ensure it's backend-storage-agnostic.

Reply:
Yes we have a lot of silently ignored failures to find and replace!

I didn't understand your rationale: "I liked keeping the option open for load_records() to be called multiple times to help ensure it's backend-storage-agnostic."

What is the use-case for calling it multiple times, what is the behaviour if you call it multiple times and how does that behaviour make it more storage-agnostic?

Reply:
Well, my thought was that there could occur situations where we wanted to pull out multiple subsets of data from a data store, or source data in batches from the underlying data store, to upload in a single job. 

If we separate out the initiation of a job, loading of records into a job, and closure of a job, we aren't bound to the assumption that we have to have all of the records present at the point that the first call to `load_records()` occurs. 

In the current implementation, if you call it multiple times, you just get multiple batches in the same job (each of will will be further chunked based on the maximum batch size).

Reply:
I've added context manager support for both query and DML operations.